### PR TITLE
ZQuantum <-> Qiskit circuit conversions for standard gates

### PR DESCRIPTION
**Pull request by alexjuda**
Implements conversion between the circuits. Includes a workaround I've came up with to overcome the circular imports between `_gates` and `_builtin_gates` modules.

Roadmap:
- ✅ zquantum built-in gates <-> Qiskit built-in gates - added in this PR
- ❌ zquantum built-in gates <-> Qiskit custom gates - not needed at the moment, all our gates are implemented in Qiskit
- ❌ zquantum custom gates <-> Qiskit built-in gates - not supported. As discussed internally, if a user stumbles upon this we should guide them to create a custom gate instead.
- 🔁 zquantum custom gates <-> Qiskit custom gates - not ready yet, will do in another PR. I've already spent so much time on this task I don't wanna risk digging to deep here without merging to dev.

### CODE REVIEW

File: src/python/zquantum/core/wip/circuit/_builtin_gates.py [hunk 1]
```diff
@@ -18,6 +18,12 @@ def builtin_gate_by_name(name) -> Optional[Union[g.Gate, GatePrototype]]:
     return globals().get(name)
 
 
+def is_non_parametric(gate_ref):
```
Comment (alexjuda) [inline src/python/zquantum/core/wip/circuit/_builtin_gates.py:21]: seems to be duplicating the one in `_gates`, is it still needed?

File: src/python/zquantum/core/wip/circuit/conversions2/qiskit_conversions.py [hunk 2]
```diff
@@ -1 +1,181 @@
...
+QISKIT_ZQUANTUM_GATE_MAP = {
+    **{q_cls: z_ref for z_ref, q_cls in ZQUANTUM_QISKIT_GATE_MAP.items()},
+    qiskit.circuit.library.CSwapGate: _make_controlled_gate_prototype(bg.SWAP),
+    qiskit.circuit.library.CRXGate: _make_controlled_gate_prototype(bg.RX),
+    qiskit.circuit.library.CRYGate: _make_controlled_gate_prototype(bg.RY),
+    qiskit.circuit.library.CRZGate: _make_controlled_gate_prototype(bg.RZ),
+}
+
+
+def convert_to_qiskit(circuit: g.Circuit) -> qiskit.QuantumCircuit:
```
Comment (alexjuda) [inline src/python/zquantum/core/wip/circuit/conversions2/qiskit_conversions.py:88]: it'd be cool to rename it to `export` to fit the `import`s below

File: src/python/zquantum/core/wip/circuit/_builtin_gates.py (hunk 1 above)
Comment (alexjuda) [reply src/python/zquantum/core/wip/circuit/_builtin_gates.py:21]: Perhaps we can check the number of `params` instead

**Comment by alexjuda** [comment]
@dexter2206 I've addressed the issues we've talked during the Zoom call. I've also added the [matrix equality checks](https://github.com/zapatacomputing/z-quantum-core/blob/4262821af10398f614a6568cbf698d610f04dcd3/src/python/zquantum/core/wip/circuit/conversions2/qiskit_conversions_test.py#L81) you'd suggested

File: src/python/zquantum/core/wip/circuit/_serde.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
+from typing import List, Iterable
+from functools import singledispatch
+
+import sympy
+
+from . import _gates as g
```
Comment (mstechly) [inline src/python/zquantum/core/wip/circuit/_serde.py:6]: Please don't use 1 and 2 letter aliases.

File: src/python/zquantum/core/wip/circuit/_serde.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+from . import _builtin_gates as bg
+from ...utils import SCHEMA_VERSION
+
+
+def serialize_expr(expr):
```
Comment (mstechly) [inline src/python/zquantum/core/wip/circuit/_serde.py:11]: It wasn't immediately obvious to me that it's about sympy expressions.
Typing or changing name would help.

**Review by dexter2206 — APPROVED** [review]

File: src/python/zquantum/core/wip/circuit/_serde.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+) -> sympy.Matrix:
+    return sympy.Matrix(
+        [
+            [deserialize_expr(element, symbols_names) for element in json_row]
+            for json_row in json_rows
+        ]
+    )
+
+
+CIRCUIT_SCHEMA = SCHEMA_VERSION + "-circuit"
```
Comment (mstechly) [inline src/python/zquantum/core/wip/circuit/_serde.py:46]: I would move this declaration at the top of the file.

File: src/python/zquantum/core/wip/circuit/_serde.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+
+def _mapv(fn, coll):
```
Comment (mstechly) [inline src/python/zquantum/core/wip/circuit/_serde.py:49]: I have no idea what this is doing and what `coll` means.

**Comment by mstechly** [comment]
@alexjuda Thanks, looks good to me 👍

**Pull request merged**